libs/vm/include/vm/module/member_function_invoke.hpp

```diff
@@ -97,34 +103,34 @@ struct MemberFunctionInvoker
     // Invoked on no parameters
     static void Invoke(VM *vm, int sp_offset, TypeId return_type_id, MemberFunction f)
     {
-      using InvokerType = MemberFunctionInvokerHelper<Type, ReturnType, MemberFunction>;
+      using InvokerType = MemberFunctionInvokerHelper<ReturnType, MemberFunction>;
       InvokerType::Invoke(vm, sp_offset, return_type_id, f);
     }
   };
 };
 
-template <typename Type, typename ReturnType, typename... Ts>
-void InvokeMemberFunction(VM *vm, TypeId return_type_id, ReturnType (Type::*f)(Ts...))
+template <typename... Args>
+struct UnpackMemberFunctionArgs
 {
-  constexpr int num_parameters         = int(sizeof...(Ts));
-  constexpr int first_parameter_offset = num_parameters - 1;
-  constexpr int sp_offset              = num_parameters;
-  using MemberFunction                 = ReturnType (Type::*)(Ts...);
-  using MemberFunctionInvoker          = typename MemberFunctionInvoker<
-      Type, ReturnType, MemberFunction>::template Invoker<first_parameter_offset, Ts...>;
-  MemberFunctionInvoker::Invoke(vm, sp_offset, return_type_id, f);
-}
+  template <typename Callable>
+  static void Invoke(VM *vm, TypeId return_type_id, Callable const &callable)
+  {
+    constexpr int num_parameters         = int(sizeof...(Args));
+    constexpr int first_parameter_offset = num_parameters - 1;
+    constexpr int sp_offset              = num_parameters;
 
-template <typename Type, typename ReturnType, typename... Ts>
-void InvokeMemberFunction(VM *vm, TypeId return_type_id, ReturnType (Type::*f)(Ts...) const)
+    using MemberFunctionInvoker =
+        typename MemberFunctionInvoker<Callable>::template Invoker<first_parameter_offset, Args...>;
+    MemberFunctionInvoker::Invoke(vm, sp_offset, return_type_id, callable);
+  }
+};
+
+template <typename Callable>
+void InvokeMemberFunction(VM *vm, TypeId return_type_id, Callable const &callable)
 {
-  constexpr int num_parameters         = int(sizeof...(Ts));
-  constexpr int first_parameter_offset = num_parameters - 1;
-  constexpr int sp_offset              = num_parameters;
-  using MemberFunction                 = ReturnType (Type::*)(Ts...) const;
-  using MemberFunctionInvoker          = typename MemberFunctionInvoker<
-      Type, ReturnType, MemberFunction>::template Invoker<first_parameter_offset, Ts...>;
```

Comment:
An example of what I meant in the PR description about passing around inferred arguments: it is redundant to pass around `Type`, `ReturnType` and `Callable`, as the former two are a property of the latter's type. `Type` is not even used until it makes it all the way from these `InvokeMemberFunction` overloads to the `InvokerHelper` at the top of this file.